contracts/governance/src/governance.rs

```diff
@@ -0,0 +1,274 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use core::iter;
+
+use alloc::vec;
+use alloc::vec::Vec;
+use canonical_derive::Canon;
+use dusk_bls12_381::BlsScalar;
+use dusk_hamt::Map;
+use dusk_jubjub::GENERATOR_EXTENDED;
+use dusk_pki::PublicKey;
+use dusk_schnorr::Signature;
+
+use crate::*;
+
+#[derive(Debug, Default, Clone, Canon)]
+pub struct GovernanceContract {
+    pub(crate) seeds: Map<BlsScalar, ()>,
+    pub(crate) balances: Map<PublicKey, u64>,
+    pub(crate) whitelist: Map<PublicKey, ()>,
+    pub(crate) paused: bool,
+    pub(crate) total_supply: u64,
+}
+
+impl GovernanceContract {
+    /// Authority of the contract
+    ///
+    /// Will have to be defined in the constant space so the bytecode of the
+    /// contract will be changed as the authority does
+    pub const AUTHORITY: PublicKey =
+        PublicKey::from_raw_unchecked(GENERATOR_EXTENDED);
+
+    fn validate_seed(
+        &mut self,
+        arguments: Vec<BlsScalar>,
+        signature: Signature,
+    ) -> Result<(), Error> {
+        let seed = arguments[0];
+
+        if self.seeds.get(&seed)?.is_some() {
+            return Err(Error::SeedAlreadyUsed);
+        }
+
+        #[cfg(target_arch = "wasm32")]
+        if !rusk_abi::verify_schnorr_sign(
+            signature,
+            Self::AUTHORITY,
+            rusk_abi::poseidon_hash(arguments),
+        ) {
+            return Err(Error::InvalidSignature);
+        }
+
+        #[cfg(not(target_arch = "wasm32"))]
+        if !signature
+            .verify(&Self::AUTHORITY, dusk_poseidon::sponge::hash(&arguments))
+        {
+            return Err(Error::InvalidSignature);
+        }
+
+        self.seeds.insert(seed, ())?;
+
+        Ok(())
+    }
+
+    fn check_pause(&self) -> Result<(), Error> {
+        (!self.paused).then_some(()).ok_or(Error::ContractIsPaused)
+    }
+
+    fn is_allowed(&self, address: &PublicKey) -> Result<(), Error> {
+        self.whitelist
+            .get(address)?
+            .as_deref()
+            .copied()
+            .ok_or(Error::AddressIsNotWhitelisted)
+    }
+
+    pub fn pause(
+        &mut self,
+        seed: BlsScalar,
+        signature: Signature,
+    ) -> Result<(), Error> {
+        self.validate_seed(
+            vec![seed, BlsScalar::from(TX_PAUSE as u64)],
+            signature,
+        )?;
+
+        self.paused = true;
+
+        Ok(())
+    }
```

Comment:
Might be irrelevant, but it is not checked whether the contract is paused already. Could that be a problem? Can something unexpected happen when a contract continuously calls `pause`?